tests/Teams/ActionBasedMessagingExtension/Bots/ActionBasedMessagingExtensionBot.cs

```diff
@@ -17,5 +18,104 @@ protected override async Task OnMessageActivityAsync(ITurnContext<IMessageActivi
         {
             await turnContext.SendActivityAsync(MessageFactory.Text($"echo: {turnContext.Activity.Text}"), cancellationToken);
         }
+
+        protected override async Task<MessagingExtensionActionResponse> OnTeamsMessagingExtensionSubmitActionAsync(ITurnContext<IInvokeActivity> turnContext, MessagingExtensionAction action, CancellationToken cancellationToken)
+        {
+            switch (action.CommandId)
+            {
+                case "createCard":
+                    return HandleCreateCardCommands(turnContext, action);
+
+                case "shareMessage":
+                    return HandleShareMessageCommand(turnContext, action);
+                default:
+                    return new MessagingExtensionActionResponse();
+            }
+        }
+
+        private MessagingExtensionActionResponse HandleCreateCardCommands(ITurnContext<IInvokeActivity> turnContext, MessagingExtensionAction action)
+        {
+            var actionData = (action.Data as JObject).ToObject<CreateCardData>();
+            var response = new MessagingExtensionActionResponse
+            {
+                ComposeExtension = new MessagingExtensionResult
+                {
+                    AttachmentLayout = "list",
+                    Type = "result",
+                },
+            };
+
+            var card = new HeroCard
+            {
+                Title = actionData.Title,
+                Subtitle = actionData.Subtitle,
+                Text = actionData.Text,
+            };
+
+            var attachments = new List<MessagingExtensionAttachment>();
+            attachments.Add(new MessagingExtensionAttachment
+            {
+                Content = card,
+                ContentType = HeroCard.ContentType,
+                Preview = card.ToAttachment(),
+            });
+
+            response.ComposeExtension.Attachments = attachments;
+            return response;
+        }
+
+        private MessagingExtensionActionResponse HandleShareMessageCommand(ITurnContext<IInvokeActivity> turnContext, MessagingExtensionAction action)
+        {
+            var heroCard = new HeroCard
+            {
+                Title = $"{action.MessagePayload.From.User.DisplayName} sent this message:",
+                Text = action.MessagePayload.Body.Content,
+            };
+            var includeImageData = (action.Data as JObject).ToObject<IncludeImageData>();
```

Comment:
Using a () style cast operator will result in an exception that the object can't be cast to this type. Using the "as" operator style cast gives you a more cryptic null exception.

Reply:
Done